frontend/src/test/scala/bloop/engine/FileWatchingSpec.scala

```diff
@@ -0,0 +1,147 @@
+package bloop.engine
+
+import java.io.{ByteArrayOutputStream, PrintStream}
+import java.nio.file.Files
+
+import bloop.Project
+import bloop.cli.{CliOptions, Commands}
+import bloop.tasks.{CompilationHelpers, ProjectHelpers}
+import bloop.tasks.ProjectHelpers.{RootProject, noPreviousResult, withState}
+import org.junit.Test
+
+class FileWatchingSpec {
+  @scala.annotation.tailrec
+  final def readCompilingLines(target: Int, msg: String, out: ByteArrayOutputStream): Int = {
+    Thread.sleep(100) // Wait 100ms for the OS's file system
+    val allContents = out.toString("UTF-8")
+    val allLines = allContents.split(System.lineSeparator())
+    val compiled = allLines.count(_.replaceAll("\u001B\\[[;\\d]*m", "").contains(msg))
+    if (compiled == target) compiled
+    else readCompilingLines(target, msg, out)
+  }
+
+  // We have here our own artificial sources
+  object ArtificialSources {
+    val `A.scala` = "package p0\ntrait A"
+    val `B.scala` = "package p1\ntrait B"
+    val `C.scala` = "package p2\nclass C extends p0.A with p1.B"
+  }
+
+  type FileWatchingContext = (State, Project, ByteArrayOutputStream)
+  def testFileWatcher(state: State, projectName: String)(
+      workerAction: FileWatchingContext => Unit,
+      testAction: (FileWatchingContext, Thread) => Unit): Unit = {
+    import scala.concurrent.Await
+    import scala.concurrent.duration
+    val projects = state.build.projects
+    val rootProject = projects
+      .find(_.name == projectName)
+      .getOrElse(sys.error(s"Project $projectName could not be found!"))
+    assert(projects.forall(p => noPreviousResult(p, state)))
+
+    // The worker thread runs the watched compilation
+    val bloopOut = new ByteArrayOutputStream()
+    val ctx: FileWatchingContext = (state, rootProject, bloopOut)
+    val workerThread = new Thread { override def run(): Unit = workerAction(ctx) }
+    implicit val context = state.executionContext
+    val testFuture = scala.concurrent.Future { testAction(ctx, workerThread) }
+    try Await.ready(testFuture, duration.Duration(15, duration.SECONDS))
+    catch { case t: Throwable => println(s"LOGS WERE ${bloopOut.toString("UTF-8")}"); throw t }
```

Comment:
Is the `println` a leftover from debugging?

Reply:
No, it's something that should help us debug in case these tests fail :smile:.